src/App.tsx

```diff
@@ -1,24 +1,278 @@
+  }, [todos]);
+
+  const clearCompletedTodos = useCallback(async () => {
+    const completedTodosIds = [...todos]
+      .filter(todo => todo.completed)
+      .map((todo) => todo.id);
+
+    setRemovingTodoIds(completedTodosIds);
+
+    Promise.all(completedTodosIds.map((id) => removeTodo(id)))
+      .then(() => {
+        setTodos(activeTodos);
+      })
+      .catch(() => {
+        setHasError(true);
+        setErrorType('delete');
+      })
+      .finally(() => setRemovingTodoIds(prevTodoIds => prevTodoIds
+        .filter((id) => !completedTodosIds.includes(id))));
+  }, [hasCompletedTodo, activeTodos]);
+
+  const handleChangeStatus = useCallback(async (todo: Todo) => {
+    setUpdatingTodoIds(prevTodoIds => [...prevTodoIds, todo.id]);
+
+    const updatedTodo = {
+      ...todo,
+      completed: !todo.completed,
+    };
+
+    try {
+      await updateTodo(todo.id, updatedTodo);
+
+      const updatedTodos = [...todos].map(todoItem => {
+        const currentTodo = todoItem;
+
+        if (currentTodo.id === updatedTodo.id) {
+          currentTodo.completed = updatedTodo.completed;
+        }
+
+        return todoItem;
+      });
+
+      setTodos(updatedTodos);
+    } catch (error) {
+      setHasError(true);
+      setErrorType('update');
+    } finally {
+      setUpdatingTodoIds(prevTodoIds => prevTodoIds
+        .filter((id) => id !== todo.id));
+    }
+  }, [activeTodos]);
+
+  const handleChangeTitle = useCallback(async (
+    todo: Todo,
+    newTitle: string,
+  ) => {
+    setUpdatingTodoIds(prevTodoIds => [...prevTodoIds, todo.id]);
+
+    const updatedTodo = {
+      ...todo,
+      title: newTitle,
+    };
+
+    try {
+      await updateTodo(todo.id, updatedTodo);
+
+      const updatedTodos = [...todos].map(todoItem => {
+        const currentTodo = todoItem;
+
+        if (currentTodo.id === updatedTodo.id) {
+          currentTodo.title = updatedTodo.title;
+        }
+
+        return todoItem;
+      });
+
+      setTodos(updatedTodos);
+    } catch (error) {
+      setHasError(true);
+      setErrorType('update');
+    } finally {
+      setUpdatingTodoIds(prevTodoIds => prevTodoIds
+        .filter((id) => id !== todo.id));
+    }
+  }, [title, updatingTodoIds, todos]);
+
+  useEffect(() => {
+    getTodosFromServer();
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const visibleTodos = todos.filter(todo => {
+    switch (selectedFilter) {
+      case FilterValues.COMPLETED:
+        return todo.completed;
+
+      case FilterValues.ACTIVE:
+        return !todo.completed;
+
+      default: return true;
+    }
+  });
+
+  const toggleAll = () => {
+    if (isEveryCompleted) {
+      const toggleToActive = [...todos].map(todoItem => {
+        if (todoItem.completed) {
+          handleChangeStatus(todoItem);
+
+          return {
+            ...todoItem,
+            completed: false,
+          };
+        }
+
+        return todoItem;
+      });
+
+      setTodos(toggleToActive);
```

Comment:
You have set changed todos inside handleChangeStatus already, so why do you do it again?)

Reply:
fixed it